src/preprocessing/passes/ackermann.cpp

```diff
@@ -187,9 +188,124 @@ void collectFunctionsAndLemmas(FunctionToArgsMap& fun_to_args,
+ * a sufficient bit-vector size. Get the converting map */
+void collectUSortsToBV(USortToBVSizeMap& usortCardinality,
+                       vector<TNode>& vec,
+                       SubstitutionMap& sortsToSkolem)
+{
+  NodeManager* nm = NodeManager::currentNM();
+
+  for (TNode term : vec)
+  {
+    if (needsReplace(term))
+    {
+      TypeNode type = term.getType();
+      size_t size = usortCardinality[type].second;
+      if (size == 0)
+      {
+        size = log2(usortCardinality[type].first) + 1;
+        usortCardinality[type].second = size;
+      }
+      Node skolem = nm->mkSkolem(
+          "BVSKOLEM$$",
+          nm->mkBitVectorType(size),
+          "a variable created by the ackermannization "
+          "preprocessing pass, representing a term with uninterpreted sort.");
+      sortsToSkolem.addSubstitution(term, skolem);
+    }
+  }
+}
+
+/* This function returns the list of terms in formula represented by assertions.
+ * We use a BFS to get all terms without duplications. */
+std::vector<TNode> getListOfTerms(AssertionPipeline* assertions)
+{
+  TNodeSet seen;
+  std::vector<TNode> res;
+  for (Node& a : assertions->ref())
+  {
+    if (seen.find(a) == seen.end())
+    {
+      res.push_back(a);
+      seen.insert(a);
+    }
+  }
+  size_t i = 0;
+  while (i < res.size())
+  {
+    for (TNode a : res[i])
+    {
+      if (seen.find(a) == seen.end())
+      {
+        res.push_back(a);
+        seen.insert(a);
+      }
+    }
+    ++i;
+  }
+  return res;
+}
+
+/* This is the top level of converting uninterpreted sorts to bit-vectors.
+ * We use BFS to get all terms without duplications, and count the number of
+ * different terms for each uninterpreted sort. Then for each sort, we will
+ * assign a new bit-vector type with a sufficient size.
+ * The size is calculated to have enough capacity, that can accommodate the
+ * variables occured in the original formula. */
+void usortsToBitVectors(const LogicInfo& d_logic,
+                        AssertionPipeline* assertions,
+                        USortToBVSizeMap& usortCardinality,
+                        SubstitutionMap& sortsToSkolem)
+{
+  std::vector<TNode> toProcess = getListOfTerms(assertions);
+  bool hasUninterpretedSorts = false;
+  for (TNode term : toProcess)
+  {
+    if (needsReplace(term))
+    {
+      hasUninterpretedSorts = true;
+      TypeNode type = term.getType();
+      /* Update the statistics for each uninterpreted sort */
+      // For non-existing keys, C++ will create a new element for it, which has
+      // the value initialized with a pair of two zeros.
+      usortCardinality[type].first = usortCardinality[type].first + 1;
+    }
+  }
+
+  if (hasUninterpretedSorts)
+  {
+    /* the current version only supports BV for removing uninterpreted sorts */
+    AlwaysAssert(d_logic.isTheoryEnabled(theory::THEORY_BV))
+        << "Cannot use Ackermannization on formula with uninterpreted "
```

Comment:
Shouldn't this change according to @mpreiner 's #3366 ?

Reply:
This change is because of been aligned with that change #3366. What's your concern here? Was I do in a wrong way?

Reply:
Looks perfectly fine ;-)

Reply:
my bad, sorry